src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -157,15 +256,299 @@ private static TensorFlowTransform Create(IHostEnvironment env, ModelLoadContext
+                fetchList.Add(args.LossOperation);
+            if (args.MetricOperation != null)
+                fetchList.Add(args.MetricOperation);
+
+            var hashedIndices = new HashSet<int>(inputColIndices);
+            for (int epoch = 0; epoch < args.Epoch; epoch++)
+            {
+                using (var cursor = input.GetRowCursor(a => hashedIndices.Contains(a)))
+                {
+                    var srcTensorGetters = GetTensorValueGetters(cursor, inputColIndices, isInputVector, tfInputTypes, tfInputShapes);
+
+                    float loss=0;
+                    float metric=0;
+                    long rows = 0;
+                    bool isDataLeft = false;
+                    using (var ch = _host.Start("Training TensorFlow model..."))
+                    using (var pch = _host.StartProgressChannel("TensorFlow training progress..."))
+                    {
+                        pch.SetHeader(new ProgressHeader("Epoch"), (e) => e.SetProgress(0, epoch));
+                        if (args.LossOperation != null)
+                            pch.SetHeader(new ProgressHeader("Average Loss"), (e) => e.SetProgress(1, loss / (rows / args.BatchSize)));
+                        if (args.MetricOperation != null)
+                            pch.SetHeader(new ProgressHeader("Average Metric"), (e) => e.SetProgress(2, metric / (rows / args.BatchSize)));
+                        while (cursor.MoveNext())
+                        {
+                            rows++;
+                            for (int i = 0; i < inputColIndices.Length; i++)
+                            {
+                                isDataLeft = true;
+                                srcTensorGetters[i].BufferTrainingData();
+                            }
+
+                            if ((rows % args.BatchSize) == 0)
+                            {
+                                isDataLeft = false;
+                                var (l, m) = TrainBatch(inputColIndices, inputsForTraining, srcTensorGetters, fetchList, args);
+                                loss += l;
+                                metric += m;
+                            }
+                        }
+                        if (isDataLeft)
+                        {
+                            isDataLeft = false;
+                            ch.Warning("Not training on the last batch. The batch size is less than {0}.", args.BatchSize);
+                        }
+                        ch.Done();
+                    }
+                }
+            }
+            UpdateModelOnDisk(model, args);
+        }
+
+        private (float loss, float metric) TrainBatch(int[] inputColIndices,
+            string[] inputsForTraining,
+            ITensorValueGetter[] srcTensorGetters,
+            List<string> fetchList,
+            Arguments args)
+        {
+            float loss = 0;
+            float metric = 0;
+            var runner = Session.GetRunner();
```

Comment:
>runner [](start = 16, length = 6)

I want to make sure I understand: the runner is tuning the weights of the graph in runner.Run()? and the weights are stored somewhere in Session?

Reply:
Yes, the line 451
> runner.AddTarget(args.OptimizationOperation);

adds the optimization operation in this runner that updates the weights. Session hold those weights.

---
In reply to: [222501506](https://github.com/dotnet/machinelearning/pull/1063#discussion_r222501506) [](ancestors = 222501506)